pkg/stub/status.go

```diff
@@ -0,0 +1,200 @@
+		},
+	}
+	err = sdk.List(h.namespace, ingressList,
+		sdk.WithListOptions(&metav1.ListOptions{}))
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf(
+			"failed to list ClusterIngresses: %v", err)
+	}
+
+	daemonsetList := &appsv1.DaemonSetList{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "DaemonSet",
+			APIVersion: "extensions/v1beta1",
+		},
+	}
+	err = sdk.List(h.namespace, daemonsetList,
+		sdk.WithListOptions(&metav1.ListOptions{}))
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf(
+			"failed to list DaemonSets: %v", err)
+	}
+
+	return ns, ingressList.Items, daemonsetList.Items, nil
+}
+
+// computeStatusConditions computes the operator's current state.
+func computeStatusConditions(conditions []osv1.ClusterOperatorStatusCondition, ns *corev1.Namespace, ingresses []ingressv1alpha1.ClusterIngress, daemonsets []appsv1.DaemonSet) []osv1.ClusterOperatorStatusCondition {
+	failingCondition := &osv1.ClusterOperatorStatusCondition{
+		Type:   osv1.OperatorFailing,
+		Status: osv1.ConditionUnknown,
+	}
+	if ns == nil {
+		failingCondition.Status = osv1.ConditionTrue
+		failingCondition.Reason = "NoNamespace"
+		failingCondition.Message = "router namespace does not exist"
+	} else {
+		failingCondition.Status = osv1.ConditionFalse
+	}
+	conditions = clusteroperator.SetStatusCondition(conditions,
+		failingCondition)
+
+	progressingCondition := &osv1.ClusterOperatorStatusCondition{
+		Type:   osv1.OperatorProgressing,
+		Status: osv1.ConditionUnknown,
+	}
+	if ingresses != nil && daemonsets != nil {
+		numIngresses := len(ingresses)
+		numDaemonsets := len(daemonsets)
+		if numIngresses == numDaemonsets {
+			progressingCondition.Status = osv1.ConditionFalse
+		} else {
+			progressingCondition.Status = osv1.ConditionTrue
+			progressingCondition.Reason = "Reconciling"
+			progressingCondition.Message = fmt.Sprintf(
+				"have %d ingresses, want %d",
+				numDaemonsets, numIngresses)
+		}
+	}
+	conditions = clusteroperator.SetStatusCondition(conditions,
+		progressingCondition)
+
+	availableCondition := &osv1.ClusterOperatorStatusCondition{
+		Type:   osv1.OperatorAvailable,
+		Status: osv1.ConditionUnknown,
+	}
+	if ingresses != nil && daemonsets != nil {
+		dsAvailable := map[string]bool{}
+		for _, ds := range daemonsets {
+			dsAvailable[ds.Name] = ds.Status.NumberAvailable > 0
+		}
+
+		unavailable := []string{}
+		for _, ingress := range ingresses {
+			name := "router-" + ingress.Name
+			if available, exists := dsAvailable[name]; !exists {
+				msg := fmt.Sprintf("no router for ingress %s",
+					ingress.Name)
+				unavailable = append(unavailable, msg)
+			} else if !available {
+				msg := fmt.Sprintf("ingress %s not available",
+					ingress.Name)
+				unavailable = append(unavailable, msg)
+			}
+		}
+
+		if len(unavailable) == 0 {
+			availableCondition.Status = osv1.ConditionTrue
+		} else {
+			availableCondition.Status = osv1.ConditionFalse
+			availableCondition.Reason = "IngressUnavailable"
```

Comment:
Could this be inverted to `IngressAvailable` to avoid double-negatives?

Reply:
The `OperatorAvailable` condition comes from the CVO API definitions, so we cannot invert it.